proxy/src/main/java/com/velocitypowered/proxy/service/VelocityServiceManager.java

```diff
@@ -0,0 +1,60 @@
+package com.velocitypowered.proxy.service;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkNotNull;
+
+import com.velocitypowered.api.service.ServiceManager;
+import com.velocitypowered.api.service.ServiceProvider;
+import com.velocitypowered.proxy.plugin.VelocityPluginManager;
+
+import java.util.Optional;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+
+public class VelocityServiceManager implements ServiceManager {
+
+  private final VelocityPluginManager pluginManager;
+  private final ConcurrentMap<Class<?>, ServiceProvider<?>>
+          registeredServices = new ConcurrentHashMap<>();
+
+  /**
+   * Initalizes the service manager.
+   *
+   * @param pluginManager the Velocity plugin manager
+   */
+  public VelocityServiceManager(VelocityPluginManager pluginManager) {
+    this.pluginManager = pluginManager;
+  }
+
+  @Override
+  public <S> boolean register(Object plugin, Class<S> serviceClass, S provider) {
+    checkNotNull(plugin, "plugin");
+    checkNotNull(serviceClass, "serviceClass");
+    checkNotNull(provider, "provider");
+    checkArgument(pluginManager.fromInstance(plugin).isPresent(), "plugin is not registered");
+    if (!registeredServices.containsKey(serviceClass)) {
+      registeredServices
+              .put(serviceClass, new VelocityServiceProvider<>(plugin, serviceClass, provider));
```

Comment:
You can use putIfAbsent for an atomic put operation, which will make the return value (very slightly) more reliable